File: matlab/problem2/runECG2_solution.m
% -------------------------------------------------------------------------
% real data, 8 blocks, more processing, feature plotting
% -------------------------------------------------------------------------

clear
close all

% sampling frequency is 512 Hz
srate = 512;

% load filters
load('filterHP_0.5_1.5.mat')
load('filterLP_3_10.mat')

% prepare figures
f1 = figure('units', 'normalized', 'outerposition', [0 0 1 1]);
f2 = figure('units', 'normalized', 'outerposition', [0 0 1 1]);
f3 = figure('units', 'normalized', 'outerposition', [0 0 1 1]);
f4 = figure('units', 'normalized', 'outerposition', [0 0 1 1]);

% phase locking pack size
packSize = 11;

% analyze blocks 1 to 8
for B=1:8
    
    %% analyze ------------------------------------------------------------
    
    % load data, get signal from channel 3
    load(['block_' num2str(B)])
    signal = block_data(:, 3);
    
    % create high-pass and band-pass filtered signals by combining filters
    signalHP = filtfilt(filterHP, 1, signal);
    signalBP = filtfilt(filterLP, 1, signalHP);
    
    % extract QRS complex positions, using the high-pass filtered signal
    [H2, HDR, s] = qrsdetect(signalHP, srate, 1);
    peakPos = H2.EVENT.POS;
    
    % compute phases
    phaseHP = angle(hilbert(signalHP));
    phaseBP = angle(hilbert(signalBP));
    
    % isolate QRS waves and phases
    timeWindow = round(-0.2*srate)+1:round(0.6*srate);
    waves = nan(length(timeWindow), length(peakPos)-2);
    phases = nan(length(timeWindow), length(peakPos)-2);
    for iEpoch=2:(length(peakPos)-1)
        wavesTmp = signalHP(peakPos(iEpoch)+(timeWindow));
        waves(:, iEpoch-1) = wavesTmp - mean(wavesTmp);
        phasesTmp = unwrap(phaseHP(peakPos(iEpoch)+timeWindow));
        phases(:, iEpoch-1) = phasesTmp - phasesTmp(timeWindow==0);
    end
    
    % compute phase locking values
    margin = floor(packSize/2);
    plv = nan(length(timeWindow), size(phases, 2));
    for iEpoch=margin+1:size(plv, 2)-margin
        thisPack = iEpoch + (-margin:margin);
        plv(:, iEpoch) = abs(mean(exp(1i*phases(:, thisPack)), 2));
    end
    
    %% plot ---------------------------------------------------------------
    
    blk = [' block ', num2str(B)];
    t = (1:length(signal))/srate;
    
    % show first 4 seconds of signal and phase
    t0 = 1;
    t1 = 4*srate;
    
    % signal detailed
    figure(f1)
    subplot(4, 2, B)
    plot(t(t0:t1), signalHP(t0:t1), 'r');
    hold on
    plot(t(t0:t1), signalBP(t0:t1), 'g');
    set(gca, 'xlim', [t(t0) t(t1)], 'ylim', [-300 300]);
    title(['detailed signal', blk])
    ylabel('\muV')
    xlabel('time [s]')
    if B==1
        legend('high-pass', 'band-pass');
    end
    
    % phase detailed
    figure(f2)
    subplot(4, 2, B)
    plot(t(t0:t1), phaseHP(t0:t1), 'r')
    hold on
    plot(t(t0:t1), phaseBP(t0:t1), 'g')
    set(gca, 'xlim', [t(t0) t(t1)], ...
        'ytick', [-pi 0 pi], ...
        'yticklabel', {'-\pi', '0', '\pi'});
    title(['detailed phase', blk])
    ylabel('\phi')
    xlabel('time [s]')
    if B==1
        legend('high-pass', 'band-pass');
    end
    
    % QRS waves
    figure(f3)
    subplot(4, 2, B)
    plot(t(1:length(timeWindow)), waves)
    hold on
    plot(t(1:length(timeWindow)), mean(waves, 2), 'linewidth', 2, ...
        'color', [0.3 0.3 0.3])
    set(gca, 'xlim', [t(1) t(length(timeWindow))], 'ylim', [-300 300]);
    title(['QRS', blk])
    ylabel('\muV')
    xlabel('time [s]')
    
    % phase locking values
    figure(f4)
    subplot(4, 2, B)
    imagesc(t(1:length(timeWindow)), margin+1:size(plv, 2)-margin, ...
        plv(:, margin+1:size(plv, 2)-margin)', [0 1])
    colormap jet
    ylabel('epoch')
    xlabel('time [s]')
    title(['phase locking', blk])
    
end

% add colorbar to PLV figure
axes, axis off
colorbar('position', [0.94 0.4 0.02 0.2])

File: python/problem2/runECG2_solution.py
"""
real data, 8 blocks, more processing, feature plotting

"""

from time import time
import sys

from pylab import *
from scipy.io import loadmat
import scipy.signal as sig

from rpeakdetect import detect_beats

rcParams.update({
    'legend.fontsize' : 'small',
    'legend.labelspacing' : 0.1,
})


# sampling frequency is 512 Hz
f_s = 512

# load filters designed with pyfda
filterHP = load('highpass_filter.npz')
filterLP = load('lowpass_filter.npz')

# phase locking pack size
packSize = 25

# analyze blocks 1 to 8
for B in range(1, 9):
    filepath = '../../data/ECG/block_%s' % B
    print('Processing block', B, end=' ')
    sys.stdout.flush()
    tic = time()

    # load data, get signal from channel 3, create time vector
    block_data = loadmat(filepath)
    block_data = block_data['block_data'].T
    ecg_signal = block_data[2]
    ecg_signal = ecg_signal.astype('float64')
    t = arange(len(ecg_signal)) / f_s

    # create high-pass and band-pass filtered signals by combining filters
    b, a = filterHP['ba']
    signalHP = sig.filtfilt(b, 1.0, ecg_signal)
    # b, a = filterLP['ba']
    signalBP = sig.filtfilt(b, 1.0, signalHP)

    # extract QRS complex positions, using the high-pass filtered signal
    peakPos = detect_beats(signalHP, f_s)

    # compute phases by getting the angle of the analytical signals
    phaseHP = angle(sig.hilbert(signalHP))
    phaseBP = angle(sig.hilbert(signalBP))

    # isolate QRS waves and phases of the high-pass filtered signal. define
    # a time window around each QRS peak, cut out the single epochs, and
    # stack them into matrices. align each signal epoch to its mean and
    # each phase epoch to its value at the time of the QRS peak. leave out
    # the first and last epoch to avoid indexing errors
    timeWindow = arange(round(-0.2 * f_s), round(0.6 * f_s))
    windows = peakPos[1:-1].reshape(-1, 1) + timeWindow
    waves = signalHP[windows]
    waves -= mean(waves, axis=0)
    phases = unwrap(phaseHP[windows])
    phases -= phases[:, timeWindow == 0]

    # compute phase locking values over time
    packNumber = len(phases) // packSize
    packs = arange(packNumber * packSize)
    packs = reshape(packs, [packNumber, packSize])
    plv = phases[packs]
    plv = abs(mean(exp(1j * plv), axis=1))

    # ----- plot ----- #
    blk = ' block %s' % B

    # show first 4 seconds of signal and phase
    t0 = 1
    t1 = int(4 * f_s)

    # signal detailed
    figure('detailed signal')
    subplot(4, 2, B)
    plot(t[t0:t1], ecg_signal[t0:t1], 'r')
    plot(t[t0:t1], signalBP[t0:t1], 'g')
    xlim([t[t0], t[t1]])
    title(blk)
    ylabel('mV')
    xlabel('time [s]')
    legend(['high-pass', 'band-pass'])

    # phase detailed
    figure('detailed phase')
    subplot(4, 2, B)
    plot(t[t0:t1], phaseHP[t0:t1], 'r')
    plot(t[t0:t1], phaseBP[t0:t1], 'g')
    xlim([t[t0], t[t1]])
    title(blk)
    ylabel('$\phi$')
    xlabel('time [s]')
    legend(['high-pass', 'band-pass'])

    # QRS waves
    figure('QRS')
    subplot(4, 2, B)
    plot(t[:len(timeWindow)], waves.T)
    plot(t[:len(timeWindow)], mean(waves, axis=0),
         linewidth=2, color=[.3, .3, .3])
    xlim([t[0], t[len(timeWindow)]])
    title(blk)
    ylabel('mV')
    xlabel('time [s]')

    # phase locking values
    figure('phase locking')
    subplot(4, 2, B)
    imshow(plv, extent=[0, 1, 0, 1], cmap='jet')
    clim([0, 1])
    ylabel('pack no.')
    xlabel('time [pt]')
    title(blk)

    toc = time()
    print('(%s sec)' % round(toc - tic, 1))

show()
